src/services/mongoose/merchants.ts

```diff
@@ -0,0 +1,166 @@
+import {
+  CouldNotFindMerchantFromIdError,
+  CouldNotFindMerchantFromUsernameError,
+} from "@domain/errors"
+
+import { Merchant } from "./schema"
+import { caseInsensitiveRegex, parseRepositoryError } from "./utils"
+
+interface IMerchantRepository {
+  listForMap(): Promise<BusinessMapMarker[] | RepositoryError>
+  listPendingApproval(): Promise<BusinessMapMarker[] | RepositoryError>
+  findById(id: MerchantId): Promise<BusinessMapMarker | RepositoryError>
+  findByUsername(username: Username): Promise<BusinessMapMarker[] | RepositoryError>
+  create(args: {
+    username: Username
+    coordinates: Coordinates
+    title: BusinessMapTitle
+    validated: boolean
+  }): Promise<BusinessMapMarker | RepositoryError>
+  update(args: {
+    id: MerchantId
+    coordinates: Coordinates
+    title: BusinessMapTitle
+    username: Username
+    validated: boolean
+  }): Promise<BusinessMapMarker | RepositoryError>
+  remove(id: MerchantId): Promise<void | RepositoryError>
+}
+
+export const MerchantsRepository = (): IMerchantRepository => {
+  const findById = async (
+    id: MerchantId,
+  ): Promise<BusinessMapMarker | RepositoryError> => {
+    try {
+      const result = await Merchant.findOne({ id })
+      if (!result) {
+        return new CouldNotFindMerchantFromIdError(id)
+      }
+      return translateToMerchant(result)
+    } catch (err) {
+      return parseRepositoryError(err)
+    }
+  }
+
+  const findByUsername = async (
+    username: Username,
+  ): Promise<BusinessMapMarker[] | RepositoryError> => {
+    try {
+      const result = await Merchant.find({ username: caseInsensitiveRegex(username) })
+      if (result.length === 0) {
+        return new CouldNotFindMerchantFromUsernameError(username)
+      }
+      return result.map(translateToMerchant)
+    } catch (err) {
+      return parseRepositoryError(err)
+    }
+  }
+
+  const listForMap = async (): Promise<BusinessMapMarker[] | RepositoryError> => {
+    try {
+      const merchants = await Merchant.find({ validated: true })
+      return merchants.map(translateToMerchant)
+    } catch (err) {
+      return parseRepositoryError(err)
+    }
+  }
+
+  const listPendingApproval = async (): Promise<
+    BusinessMapMarker[] | RepositoryError
+  > => {
+    try {
+      const merchants = await Merchant.find({ validated: false })
+      return merchants.map(translateToMerchant)
+    } catch (err) {
+      return parseRepositoryError(err)
+    }
+  }
+
+  const create = async ({
+    username,
+    coordinates,
+    title,
+    validated,
+  }: {
+    username: Username
+    coordinates: Coordinates
+    title: BusinessMapTitle
+    validated: boolean
+  }): Promise<BusinessMapMarker | RepositoryError> => {
+    try {
+      const location = {
```

Comment:
Let's define this as our type.

Reply:
you mean Promise<BusinessMapMarker | RepositoryError> ?